### PR TITLE
Exposing Serve Service

### PR DESCRIPTION
**Pull request by kodwanis**
## Why are these changes needed?
This PR exposes the serve service and all of its fields as part of the RayService CRD. This allow users to pass custom fields like labels, annotations and other fields to Serve Service. 

This is similar to exposing head service to users addressed in https://github.com/ray-project/kuberay/pull/1040

**Here's the definitions of the priorities:**

**Name**: If it's specified in ServeService, override the default.

**Namespace**: ignore user-provided namespace in ServeService. Enforce it to be the same namespace as the RayCluster.

**annotations field**: user-provided fields are passed as is.

**Selector field**: Ignore user specified selector fields and keep the default fields

**Labels field**:  default field values takes priority user specified fields. Other user specified fields passed are merged.

**Ports field**: Keeping the behavior consisted of adding only serve port. If serve port is defined in rayCluster, that takes priority and user specified ports are ignored. If serve port is not defined in rayCluster, user specified serve port is added. All ports other than serve port are ignored.

**Type**: If it's specified in ServeService, override the default.

## Expected result
The `serve-svc` should have the service defined by the user. It should contain the user defined labels, annotations, Type, name (if passed) and "serve" port (if not defined in the cluster config).

Expected output for [ray-service.custom-serve-service.yaml](https://github.com/ray-project/kuberay/pull/1117/files#diff-27108740a922d18248ded5e6295ba2ff512afd6106d7280dd1ca31dfe8f90412):
```yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: 'XX'
  labels:
    ray.io/serve: custom-ray-serve-service-name-serve
    ray.io/service: custom-ray-serve-service-name
    custom-label: custom-ray-serve-service-label
  annotations:
    custom-annotation: custom-ray-serve-service-annotation
  name: custom-ray-serve-service-name-serve-svc
  namespace: default
  ownerReferences:
    - apiVersion: ray.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: RayService
      name: custom-ray-serve-service-name-
      uid: xxx
  resourceVersion: '3265983567'
  uid: xxx
spec:
  clusterIP: 172.30.233.159
  clusterIPs:
    - 172.30.233.159
  internalTrafficPolicy: Cluster
  ipFamilies:
    - IPv4
  ipFamilyPolicy: SingleStack
  ports:
    - name: serve
      port: 8000
      protocol: TCP
      targetPort: 8000
  selector:
    ray.io/cluster: custom-ray-serve-service-name-serve-raycluster-xxxx
    ray.io/serve: 'true'
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer: {}
```

## Related issue number

Closes https://github.com/ray-project/kuberay/issues/1034

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by architkulkarni — COMMENTED** [review]
It looks like there's a fair amount of duplicated code and test code from https://github.com/ray-project/kuberay/pull/1040/, is it possible to refactor it to avoid duplicating code as much as possible?  For example, by pulling out common code into helper functions.  

For the test code, I'm less sure, but I think it's possible to use "table-driven tests" to parametrize multiple tests that are very similar.

Will give this a detailed review as soon as I can.

**Comment by kodwanis** [comment]
> It looks like there's a fair amount of duplicated code and test code from #1040, is it possible to refactor it to avoid duplicating code as much as possible? For example, by pulling out common code into helper functions.
> 
> For the test code, I'm less sure, but I think it's possible to use "table-driven tests" to parametrize multiple tests that are very similar.
> 
> Will give this a detailed review as soon as I can.

@architkulkarni Added helper function to reduce the duplicated code both in the implementation and test files. Please take a look.

File: ray-operator/controllers/ray/common/service.go
Comment (architkulkarni) [inline ray-operator/controllers/ray/common/service.go:178]:
```suggestion
		// Add DefaultServePort if it is already not added and ignore any custom ports
		// Keeping this consistent with adding only serve port in serve service
```

Comment (architkulkarni) [reply ray-operator/controllers/ray/common/service.go:178]: Does it make sense to print or log a warning when we see a custom port and ignore it?

**Review by architkulkarni — APPROVED** [review]
Thanks for the refactor, looks good!  Just one nit about the warning but it shouldn't block the merge. 

Before merging I would like to get @kevin85421's opinion on just one point, which is the handling of ports. @kevin85421 do you think it makes sense? I don't have any strong preferences here for how to handle them.

File: ray-operator/controllers/ray/common/service.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service.go:178]: I can add the warning.

**Comment by architkulkarni** [comment]
@kodwanis Sorry, do you mind fixing the last merge conflict?

File: ray-operator/controllers/ray/common/service_test.go
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service_test.go:428]: `validateNameAndNamespaceForUserSpecifiedService`?

File: ray-operator/controllers/ray/common/service_test.go
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service_test.go:497]: Should we also check the value of the port?

**Review by kevin85421 — COMMENTED** [review]
Would you mind adding the expected result of [ray-service.custom-serve-service.yaml](https://github.com/ray-project/kuberay/pull/1117/files#diff-27108740a922d18248ded5e6295ba2ff512afd6106d7280dd1ca31dfe8f90412) in the PR description? I will clone this branch and manually test it. Thanks!

File: ray-operator/controllers/ray/common/service_test.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service_test.go:489]: Added a check for value of the port

File: ray-operator/controllers/ray/common/service_test.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service_test.go]: Updated

File: ray-operator/controllers/ray/common/service.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service.go:187]: Updated the comment to be more clear that only port with name "serve" will be added

File: ray-operator/controllers/ray/common/service.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service.go:93]: Ah, this must be from the resolving conflict. 
Moved to nil check to `setLabelsforUserProvidedService` and removed the redundant code

**Comment by kodwanis** [comment]
> Would you mind adding the expected result of [ray-service.custom-serve-service.yaml](https://github.com/ray-project/kuberay/pull/1117/files#diff-27108740a922d18248ded5e6295ba2ff512afd6106d7280dd1ca31dfe8f90412) in the PR description? I will clone this branch and manually test it. Thanks!

@kevin85421 Updated the PR description with expected output for `serve-svc`

**Review by kevin85421 — APPROVED** [review]
Others look good to me.

File: ray-operator/controllers/ray/common/service.go
Comment (kodwanis) [reply ray-operator/controllers/ray/common/service.go]: Updated the log statement

**Comment by kevin85421** [comment]
I tested it manually, comparing the results with and without this PR. The upper one corresponds to the results with this PR applied, while the lower one represents the results without this PR. I will attempt to rerun the failed nightly compatibility tests. However, the nightly tests have become quite unstable recently, possibly due to breaking changes in the Ray master branch. If it still cannot pass the test, we may consider merging this PR without passing that test.

Screen Shot 2023-06-06 at 2 59 44 PM

**Comment by Jeffwan** [comment]
Hi, I suggest to hold this change and introduce to beta APIs.

**Comment by kodwanis** [comment]
> Hi, I suggest to hold this change and introduce to beta APIs. #1146 is added and we can grow new fields in latest API version and controller version.

@Jeffwan This has been a critical blocker for adoption for months now and this is a must have for load balance the traffic for rayserve.

**Comment by kevin85421** [comment]
Hi @Jeffwan, we may not update the version of RayService for v0.6.0. Instead, we will update the versions for RayJob and RayCluster in v0.6.0. I believe it is safe to merge this PR as it doesn't introduce any backward compatibility issues or cause significant delays for users. This PR is similar to Archit's PR #1040. Is it OK for you? Thanks!

**Comment by kevin85421** [comment]
Experiments:

```sh
# Run test
RAY_IMAGE=rayproject/ray:nightly OPERATOR_IMAGE=controller:latest python3 tests/compatibility-test.py RayFTTestCase.test_ray_serve 2>&1 | tee log

# Check Ray commit (in head Pod)
python3 -c "import ray; print(ray.__commit__)"
```
* KubeRay: 2de3fe5ca3cf206c4ebb9912e128295e5cc5db45, Ray: [2081ebb5724785f60a77b4ec482fb7e8f9baab0d](https://github.com/ray-project/ray/pull/36148) => I tried twice: 1 Fail and 1 Pass
* KubeRay: v0.5.0, Ray: [2081ebb5724785f60a77b4ec482fb7e8f9baab0d](https://github.com/ray-project/ray/pull/36148) => 1 Pass

**Review by kevin85421 — APPROVED** [review]
I am not sure why does it consistently fails on GitHub Actions, but it can pass on my devbox. We need to prioritize #1058.

Screen Shot 2023-06-07 at 5 10 25 PM

Experiments:

```sh
# Run test
RAY_IMAGE=rayproject/ray:nightly OPERATOR_IMAGE=controller:latest python3 tests/compatibility-test.py RayFTTestCase.test_ray_serve 2>&1 | tee log

# Check Ray commit (in head Pod)
python3 -c "import ray; print(ray.__commit__)"
```
* KubeRay: This PR, Ray: [2081ebb5724785f60a77b4ec482fb7e8f9baab0d](https://github.com/ray-project/ray/pull/36148) => 1 Pass

**Comment by kevin85421** [comment]
Merge this PR before passing the nightly compatibility test. The failing has no relationship with this PR. See https://github.com/ray-project/kuberay/pull/1117#pullrequestreview-1468758930 for more details.

**Pull request merged**
